File: leetcode-cn/Python/38.外观数列.py
#
# @lc app=leetcode.cn id=38 lang=python3
#
# [38] 外观数列
#


# @lc code=start
class Solution:
    def countAndSay(self, n: int) -> str:
        if n == 1:
            return '1'

        s = self.countAndSay(n - 1)

        i = 0
        res = ''

        for j, c in enumerate(s):
            if c != s[i]:
                res += str(j - i) + s[i]
                i = j

        res += str(len(s) - i) + s[-1]
        return res


# @lc code=end

File: leetcode-cn/C++/38.外观数列.cpp
/*
 * @lc app=leetcode.cn id=38 lang=cpp
 *
 * [38] 外观数列
 */

// @lc code=start
#include <string>
using namespace std;

class Solution {
public:
    string countAndSay(int n)
    {
        if (n == 1) {
            return "1";
        }

        string preRet = countAndSay(n - 1);
        string ans = "";
        char pre = preRet[0];
        int cnt = 1;
        for (int i = 1; i < preRet.size(); i++) {
            if (preRet[i] == preRet[i - 1]) {
                cnt += 1;
            } else {
                ans = ans + to_string(cnt) + pre;
                pre = preRet[i];
                cnt = 1;
            }
        }

        ans = ans + to_string(cnt) + pre;
        return ans;
    }
};
// @lc code=end
